src/index.ts

```diff
@@ -153,29 +153,67 @@ const getAssetFromKV = async (event: FetchEvent, options?: Partial<Options>): Pr
 
   if (response) {
     let headers = new Headers(response.headers)
-    headers.set('CF-Cache-Status', 'HIT')
     if (shouldSetBrowserCache) {
       headers.set('cache-control', `max-age=${options.cacheControl.browserTTL}`)
     } else {
       // don't assume we want same cache behavior of edge TTL on client
       // so remove the header from the response we'll return
       headers.delete('cache-control')
     }
-    response = new Response(response.body, { headers })
+
+    let shouldRevalidate = false
+    // Four preconditions must be met for a 304 Not Modified:
+    // - the request cannot be a range request
+    // - client sends if-none-match
+    // - resource has etag
+    // - test if-none-match against the pathKey so that we test against KV, rather than against
+    // CF cache, which may modify the etag with a weak validator (e.g. W/"...")
+    shouldRevalidate = [
+      request.headers.has('range') !== true,
+      request.headers.has('if-none-match'),
+      response.headers.has('etag'),
+      request.headers.get('if-none-match') === `${pathKey}`,
+    ].every((val) => val === true)
+
+    if (shouldRevalidate) {
+      // TypeError permitted to satisfy mocks per /issues/96
+      try {
+        // Proactively remove body per /pull/94#discussion_r425455176
+        response.body.cancel()
+      } catch (e) {
+        if (e instanceof TypeError && e.message === 'response.body.cancel is not a function') {
+          console.log('Environment doesnt support readable streams')
+        }
+      }
+
+      headers.set('cf-cache-status', 'REVALIDATED')
+      response = new Response(null, {
+        status: 304,
+        headers,
+        statusText: 'Not Modified',
+      })
+    } else {
+      headers.set('CF-Cache-Status', 'HIT')
+      response = new Response(response.body, { headers })
+    }
   } else {
     const body = await ASSET_NAMESPACE.get(pathKey, 'arrayBuffer')
     if (body === null) {
       throw new NotFoundError(`could not find ${pathKey} in your content namespace`)
     }
     response = new Response(body)
 
-    // TODO: could implement CF-Cache-Status REVALIDATE if path w/o hash existed in manifest
-
     if (shouldEdgeCache) {
-      response.headers.set('CF-Cache-Status', 'MISS')
+      response.headers.set('Accept-Ranges', 'bytes')
+      response.headers.set('Content-Length', body.length)
+      // set etag before cache insertion
+      if (!response.headers.has('etag')) {
+        response.headers.set('etag', `${pathKey}`)
```

Comment:
```suggestion
        response.headers.set('etag', `W/"${pathKey}"`)
```

Use [Weak ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) for performance, See https://support.cloudflare.com/hc/en-us/articles/218505467-Using-ETag-Headers-with-Cloudflare

Reply:
Nice! But does this matter though since the `etag` is being so explicitly set on each KV asset?

Each new or updated asset always has a new hash. There would no strong/weak pairing in this scenario?

Reply:
Hey @SukkaW - thanks for your input. Weakly held etags are used (by us at Cloudflare at least), when a response from a remote origin server is modified by an intermediate proxy before being served to the client. For example, if an origin server sends a gzip-compressed file, but Cloudflare applies brotli compression to the resource before sending it to the client, we will use the weak etag comparison algorithm. In other words, Cloudflare is saying to the client "this resource is nearly identical to the resource we got from the origin, but its not the original resource, so we're going to change the etag so no one gets confused"

In this scenario as @whatnickcodes mentioned, Cloudflare is acting as the origin server and there is no intermediate proxy. As such, there is only one version of the etag and one version of the resource. Applying a weak etag would be redundant in this case :)

Reply:
Save the byte!

Reply:
@shagamemnon Thanks for the explanation! I will mark this comment as resoloved.